File: raspberry-pi-ios-client/RaspberryPiClient/Presentation/Connection/ConnectionViewModel.swift
//
//  ConnectionViewModel.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/10/26.
//

import Combine
import Foundation

protocol ConnectionViewModelProtocol: ObservableObject {
    var ipAddress: String { get set }
    var connectionMode: ConnectionMode { get set }
    var connectionStatus: WebSocketConnectionState { get }
    var isConnectButtonDisabled: Bool { get }
    func connect()
    func disconnect()
}

class ConnectionViewModel: ConnectionViewModelProtocol {
    @Published private(set) var connectionStatus = WebSocketConnectionState.ready
    @Published var ipAddress: String = ""
    @Published var connectionMode: ConnectionMode = .rosBridge

    private var observationTask: Task<Void, Never>?
    private let rosBridgeConnectionRepository: RosBridgeConnectionRepositoryProtocol

    init(rosBridgeConnectionRepository: RosBridgeConnectionRepositoryProtocol = RosBridgeConnectionRepository()) {
        self.rosBridgeConnectionRepository = rosBridgeConnectionRepository
        self.startObservingConnectionState()
    }
    
    deinit {
        observationTask?.cancel()
    }
    
    var isConnectButtonDisabled: Bool {
        ipAddress.trimmingCharacters(in: .whitespaces).isEmpty || connectionStatus == .connected
    }

    private func startObservingConnectionState() {
        observationTask = Task { @MainActor in
            for await state in rosBridgeConnectionRepository.observeConnectionState() {
                self.connectionStatus = state
            }
        }
    }

    func connect() {
        rosBridgeConnectionRepository.connect(ipAddress: ipAddress, connectionMode: connectionMode)
    }
    
    func disconnect() {
        rosBridgeConnectionRepository.disconnect()
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_android_client/presentation/screen/connection/ConnectionViewModel.kt
package com.shusuke.raspberry_pi_android_client.presentation.screen.connection

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.shusuke.raspberry_pi_android_client.data.infrastructure.websocket.WebSocketConnectionState
import com.shusuke.raspberry_pi_android_client.data.repository.connection.ConnectionMode
import com.shusuke.raspberry_pi_android_client.data.repository.connection.RosConnectionRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/**
 * 接続画面用 ViewModel（iOS [ConnectionViewModel] に相当）。
 */
class ConnectionViewModel(
    private val rosConnectionRepository: RosConnectionRepository,
) : ViewModel() {

    private val _connectionStatus =
        MutableStateFlow<WebSocketConnectionState>(WebSocketConnectionState.Ready)
    val connectionStatus: StateFlow<WebSocketConnectionState> = _connectionStatus.asStateFlow()

    private val _ipAddress = MutableStateFlow("")
    val ipAddress: StateFlow<String> = _ipAddress.asStateFlow()

    private val _connectionMode = MutableStateFlow(ConnectionMode.ROS_BRIDGE)
    val connectionMode: StateFlow<ConnectionMode> = _connectionMode.asStateFlow()

    val isConnectButtonDisabled: StateFlow<Boolean> = combine(
        _ipAddress,
        _connectionStatus,
    ) { ip, status ->
        ip.trim().isBlank() ||
            status is WebSocketConnectionState.Connecting ||
            status is WebSocketConnectionState.Connected
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = true,
    )

    init {
        viewModelScope.launch {
            rosConnectionRepository.observeConnectionState().collect { state ->
                _connectionStatus.value = state
            }
        }
    }

    fun setIpAddress(value: String) {
        _ipAddress.value = value
    }

    fun setConnectionMode(mode: ConnectionMode) {
        _connectionMode.value = mode
    }

    fun connect() {
        rosConnectionRepository.connect(
            ipAddress = _ipAddress.value.trim(),
            connectionMode = _connectionMode.value,
        )
    }

    fun disconnect() {
        rosConnectionRepository.disconnect()
    }
}
